### ISSUE TITLE
restic serve: Didn't finish writing GET request

### ISSUE THREAD

**Issue by bkmeneguello** [rclone/rclone#2598]
I'm getting this message sometimes. When this error occurs, shouldn't the operation retry? If the copy doesn't succeed the file is left corrupt?

https://github.com/rclone/rclone/blob/9fa8c959ee331938a8a05fbe318ce61af1296972/cmd/serve/restic/restic.go#L315

**Comment by ncw** [comment]
This is restic fetching an object from an rclone backend.

rclone writes the HTTP header and then copies the content from the backend to restic as the HTTP response body.

If there is an error in that process - let's say some sort of network problem - rclone will print that message and finish the transfer early.  Rclone has already sent the headers at that point so it can't send an errored HTTP response.

However rclone writes a `Content-Length` header in the response, so restic should know the blob it was fetching is too short and try again - is that right @fd0 ?

It is possible that restic just didn't read all the data in the blob and closed the connection early - it is quite entitled to do that.

Do  you have an example of the actual log message?

(Looking at the code, this bit seems to be left over and should probably be deleted...

https://github.com/ncw/rclone/blob/9fa8c959ee331938a8a05fbe318ce61af1296972/cmd/serve/restic/restic.go#L255-L256

As the `Content-Length` is set properly here after a potential Range request has been decoded

https://github.com/ncw/rclone/blob/9fa8c959ee331938a8a05fbe318ce61af1296972/cmd/serve/restic/restic.go#L286

)

**Comment by bkmeneguello** [comment]
I don't have the full message anymore. I was restoring my backup using rclone backend and be worried about these messages. Since it's a critical backup I aborted and used the native restic s3 backend.
I prefer to use rclone, but I need to have sure all files are completely restored.
I'll try to restore another piece and if this error happens again I'll add it here.

**Comment by fd0** [comment]
> However rclone writes a Content-Length header in the response, so restic should know the blob it was fetching is too short and try again - is that right @fd0 ?

Yes, all errors are retried by restic. It would be interesting to know which restic command you ran, and what restic printed...

**Comment by tipruzs** [comment]
@fd0 similar issue here. i'm using rclone as restic backend and a sftp target for rclone.

```bash
restic --cache-dir /tmp -o rclone.args="serve restic --stdio" -r "rclone:restic:/path/to/data" restore -t /restore <snapshot id>
```

```
rclone: 2018/10/15 10:40:24 ERROR : data/05/057ac743e41d3e71da910ee334126374c5b3a578e6f7961761971dd5b928ce68: Didn't finish writing GET request (wrote 262144/272009 bytes): http2: stream closed
rclone: 2018/10/15 10:40:29 ERROR : data/9e/9e0eea7995a911b11e3d81c52638fda6f9288165d970f2242d8a539e18d84cc3: Didn't finish writing GET request (wrote 32768/41197 bytes): http2: stream closed
rclone: 2018/10/15 10:41:24 ERROR : data/9e/9e0eea7995a911b11e3d81c52638fda6f9288165d970f2242d8a539e18d84cc3: Didn't finish writing GET request (wrote 262144/290432 bytes): http2: stream closed
rclone: 2018/10/15 10:44:18 ERROR : data/7c/7ca5cc5d21ba93127473dfed4bd1bbbd3e0b48884d148413b7f5f0bbb5b336a0: Didn't finish writing GET request (wrote 0/4679 bytes): http2: stream closed
rclone: 2018/10/15 10:45:17 ERROR : data/02/028df112560b8d3936035e1fac65c0bbbcaf953ba0bf71587268bc530619b1f9: Didn't finish writing GET request (wrote 0/7482 bytes): http2: stream closed
rclone: 2018/10/15 10:45:37 ERROR : data/02/028df112560b8d3936035e1fac65c0bbbcaf953ba0bf71587268bc530619b1f9: Didn't finish writing GET request (wrote 0/7918 bytes): http2: stream closed
rclone: 2018/10/15 10:50:04 ERROR : data/7c/7ca5cc5d21ba93127473dfed4bd1bbbd3e0b48884d148413b7f5f0bbb5b336a0: Didn't finish writing GET request (wrote 0/4280 bytes): http2: stream closed
rclone: 2018/10/15 10:50:12 ERROR : data/7c/7ca5cc5d21ba93127473dfed4bd1bbbd3e0b48884d148413b7f5f0bbb5b336a0: Didn't finish writing GET request (wrote 0/4239 bytes): http2: stream closed
rclone: 2018/10/15 10:51:25 ERROR : data/35/35a83ee95ecb1808381bcbe254b7c329e23b538933f46c1eb6401eb3d411796f: Didn't finish writing GET request (wrote 32768/41458 bytes): http2: stream closed
rclone: 2018/10/15 10:52:51 ERROR : data/f2/f2955e493f2571985d183b8af7ec05fb7e1958a659ea257fe252b5a30caaf3ea: Didn't finish writing GET request (wrote 32768/40679 bytes): http2: stream closed
rclone: 2018/10/15 10:58:09 ERROR : data/d3/d341034f3d7eae87a84a205d4015de6a910707b7c22e07ab8ad37b535d85c6ce: Didn't finish writing GET request (wrote 0/7200 bytes): http2: stream closed
rclone: 2018/10/15 11:01:14 ERROR : data/35/3549fa628401dd79856e15e5daf9931d2c9669910867035915d558d029d32575: Didn't finish writing GET request (wrote 0/4925 bytes): http2: stream closed
rclone: 2018/10/15 11:02:04 ERROR : data/35/3549fa628401dd79856e15e5daf9931d2c9669910867035915d558d029d32575: Didn't finish writing GET request (wrote 0/5113 bytes): http2: stream closed
rclone: 2018/10/15 11:03:59 ERROR : data/97/971bf1c28fda3c7fbe775dc6d69d356bc49b054778950a1494f052e5d84a8db1: Didn't finish writing GET request (wrote 32768/46454 bytes): http2: stream closed
rclone: 2018/10/15 11:04:45 ERROR : data/35/3549fa628401dd79856e15e5daf9931d2c9669910867035915d558d029d32575: Didn't finish writing GET request (wrote 65536/72370 bytes): http2: stream closed
```
Do you need any further information?

**Comment by ncw** [comment]
@tipruzs are there any restic logs matching those logs from rclone?

**Comment by tipruzs** [comment]
@ncw did not see any relevant logline from restic so i restarted the restore with `--verbose`. But still no hint whats going wrong.

```
restic --verbose --verbose --cache-dir /tmp -o rclone.args="${RCLONE_ARGS}" -r "${RESTIC_REPOSITORY}" restore -t /restore <snapshot-id>
rclone: 2018/10/15 11:33:25 NOTICE: Config file "/root/.config/rclone/rclone.conf" not found - using defaults
repository 1740955a opened successfully, password is correct
restoring <Snapshot <snapshot-id> of [/source] at 2018-10-15 05:35:39.865446561 +0200 CEST by @backup-job> to /restore
rclone: 2018/10/15 11:35:43 ERROR : data/05/057ac743e41d3e71da910ee334126374c5b3a578e6f7961761971dd5b928ce68: Didn't finish writing GET request (wrote 32768/40866 bytes): http2: stream closed
```

**Comment by ncw** [comment]
@tipruzs as I said above it isn't really an error, it is just rclone reporting that it didn't send all the data it was expecting to.

restic might have had a good reason to close the file early before reading to the end - I don't know whether restic does that. @fd0 ?  If that is something restic normally does then I should just downgrade the error to a debug.

**Comment by fd0** [comment]
Hm, restic should not do that. @tipruzs what does `restic version` report? Can you try to reproduce this issue (ideally with instructions) with a fresh repo and just one or two files in the backup?

**Comment by tipruzs** [comment]
@fd0 i'm running latest stable: `restic 0.9.3 compiled with go1.11.1 on linux/amd64`
I need some time to prepare your request.

**Comment by tipruzs** [comment]
@fd0 here it is. i created the following poc for this issue:

Sample Data located at logical path **/poc/source**:
- https://wordpress.org/latest.zip (extracted, many small files)
- https://www.sample-videos.com/video/mp4/720/big_buck_bunny_720p_30mb.mp4 (one large file)

Created a dummy mega account for rclone and reuse it in the following env vars:
```
RESTIC_REPOSITORY=rclone:restic:/
RESTIC_PASSWORD=0123456789
RCLONE_ARGS=serve restic --stdio
RCLONE_CONFIG_RESTIC_TYPE=mega
RCLONE_CONFIG_RESTIC_USER=rclone-issues-2598@mailinator.com
RCLONE_CONFIG_RESTIC_PASS=Cpocn64yQJHI45T-Afb4uaVtXtHENOq6dJs
```

Created a new repository and started initial backup
```
restic --cache-dir /tmp -o rclone.args="${RCLONE_ARGS}" -r "${RESTIC_REPOSITORY}" init
restic --cache-dir /tmp -o rclone.args="${RCLONE_ARGS}" -r "${RESTIC_REPOSITORY}" backup --hostname poc-2598 /poc/source
```

Restore files from this snapshot to **/poc/restored**:
```
restic --cache-dir /tmp -o rclone.args="${RCLONE_ARGS}" -r "${RESTIC_REPOSITORY}" restore -t /poc/restored 1ee3e450
rclone: 2018/10/16 10:29:36 NOTICE: Config file "/root/.config/rclone/rclone.conf" not found - using defaults
repository 9793ee84 opened successfully, password is correct
restoring <Snapshot 1ee3e450 of [/poc/source] at 2018-10-16 10:28:00.920943931 +0200 CEST by @poc-2598> to /poc/restored
rclone: 2018/10/16 10:32:53 ERROR : data/62/626dd830d5be31c86ad8291925f7d2ab32bc65316eb6b9c4d3a73c74086284e3: Didn't finish writing GET request (wrote 0/4269 bytes): http2: stream closed
```

**Comment by ncw** [comment]
I managed to replicate the problem on the first run using a swift backend, but it never did it again after that!

**Comment by tipruzs** [comment]
@ncw i can confirm, that the frequency of the error message is much higher with a sftp backend (e.g. Strato Hidrive). Unfortunately i found no public sftp service to prepare a poc for you.

**Comment by fd0** [comment]
I've discovered that restic's debug log does not log HTTP requests sent to rclone, I've fixed that in master. Afterwards, I unfortunately did not have time any more to dig into this further today.

**Comment by azurefreecovid** [comment]
I'm also getting these error messages using Restic with Rclone pointing to an SFTP backend. The client (where Restic and Rclone are running) is Windows 10.

**Comment by ivoanjo** [comment]
I'm affected by this while running `restic check --read-data` with rclone+jottacloud (I also replied in restic/restic#2742 ). I've ran into the issue with a debug build of restic, here's what I got in my logs @fd0 :

Restic output logs:

```
rclone: 2020/05/29 23:49:55 ERROR : data/df/df4d26fdddea9f0448853119f3a9cb52c82d6ce1ef0ae4f8c6b8b82cdb8eeb2f: Didn't finish writing GET request (wrote 0/4995802 bytes): unexpected EOF
Pack ID does not match, want df4d26fd, got e3b0c442
```

restic debug logs:

```
2020/05/29 23:49:49 checker/checker.go:699      checker.checkPack       973     checking pack df4d26fdddea9f0448853119f3a9cb52c82d6ce1ef0ae4f8c6b8b82cdb8eeb2f
2020/05/29 23:49:49 cache/backend.go:196        cache.(*Backend).Load   973     Load(<data/df4d26fddd>, 0, 0): delegating to backend
2020/05/29 23:49:49 rest/rest.go:180    rest.(*Backend).openReader      973     Load <data/df4d26fddd>, length 0, offset 0
2020/05/29 23:49:49 rest/rest.go:204    rest.(*Backend).openReader      973     Load(<data/df4d26fddd>) send range bytes=0-
2020/05/29 23:49:49 debug/round_tripper_debug.go:77     debug.loggingRoundTripper.RoundTrip     973     ------------  HTTP REQUEST -----------
GET /data/df4d26fdddea9f0448853119f3a9cb52c82d6ce1ef0ae4f8c6b8b82cdb8eeb2f HTTP/1.1
Host: localhost
User-Agent: Go-http-client/1.1
Accept: application/vnd.x.restic.rest.v2
Range: bytes=0-
2020/05/29 23:49:55 debug/round_tripper_debug.go:90     debug.loggingRoundTripper.RoundTrip     973     ------------  HTTP RESPONSE ----------
HTTP/2.0 206 Partial Content
Accept-Ranges: bytes
Content-Range: bytes 0-4995801/4995802
Date: Fri, 29 May 2020 22:49:55 GMT
Server: rclone/v1.51.0
Content-Length: 0

2020/05/29 23:49:55 checker/checker.go:712      checker.checkPack       973     hash for pack df4d26fdddea9f0448853119f3a9cb52c82d6ce1ef0ae4f8c6b8b82cdb8eeb2f is e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
2020/05/29 23:49:55 checker/checker.go:715      checker.checkPack       973     Pack ID does not match, want df4d26fdddea9f0448853119f3a9cb52c82d6ce1ef0ae4f8c6b8b82cdb8eeb2f, got e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
```

Versions:

* `restic 0.9.6 (v0.9.6-207-g84475aa3) compiled with go1.14.3 on linux/amd64`
*
  ```
  rclone v1.51.0
  - os/arch: linux/amd64
  - go version: go1.13.7
  ```

Let me know if I can provide any more info.

**Comment by ncw** [comment]
> **Edit**: Hmm, looking at the logs, it looks like rclone is setting `Content-Length: 0` in the response.

That **is** interesting...

This looks like it is the cause of the problem. I can't see how rclone can be setting the Content-Length to 0 though - check the code here. We know what quite a few of the values must be from the Content-Range header.

https://github.com/rclone/rclone/blob/e7bd392a69bda63d84ca1ccdb8a97cf571feefba/cmd/serve/httplib/serve/serve.go#L44-L70

But if somehow restic thinks it is 0 then it would fit the error message that rclone gives perfectly "Didn't finish writing GET request (wrote 0/4995802 bytes)" - rclone thinks it has to write 4995802 bytes but restic only reads 0 bytes.

It would be great to work out if rclone is sending `Content-Length: 0` or not or whether seeing it as 0 is an artifact of the go http libraries or something that in restic.

This could be done by setting `rclone restic serve` up to serve from a local IP address rather than over a socket, then using wireshark to sniff the connection.

This could also be a bug in the go http2 handling which rclone uses by default...

It might be worth trying rclone v1.52 which was just released to see if you get any different results as that uses a newer go compiler and standard library.

**Comment by ivoanjo** [comment]
I've been staring at that file and noticed one more interesting thing:

https://github.com/rclone/rclone/blob/e7bd392a69bda63d84ca1ccdb8a97cf571feefba/cmd/serve/httplib/serve/serve.go#L70-L88

The `Content-Length` gets set from the `size`, but the same `size` gets used in the error message: `rclone: 2020/05/29 23:49:55 ERROR : data/df/df4d26fdddea9f0448853119f3a9cb52c82d6ce1ef0ae4f8c6b8b82cdb8eeb2f: Didn't finish writing GET request (wrote 0/4995802 bytes): unexpected EOF`.

Thus, it doesn't look like that code is setting the wrong size -- otherwise we'd see it in the message. So either the wrong size gets set elsewhere in the server, or it's misinterpreted by restic.

One extra weird detail, that may or may not be relevant is the actual position of the header in the logs...

Consider (the failed request)

```
2020/05/29 23:49:55 debug/round_tripper_debug.go:90     debug.loggingRoundTripper.RoundTrip     973     ------------  HTTP RESPONSE ----------
HTTP/2.0 206 Partial Content
Accept-Ranges: bytes
Content-Range: bytes 0-4995801/4995802
Date: Fri, 29 May 2020 22:49:55 GMT
Server: rclone/v1.51.0
Content-Length: 0
```

versus a request that goes well:

```
HTTP/2.0 206 Partial Content
Content-Length: 4433053
Accept-Ranges: bytes
Content-Range: bytes 0-4433052/4433053
Content-Type: application/octet-stream
Date: Fri, 29 May 2020 22:49:50 GMT
Server: rclone/v1.51.0
```

The `Content-Length` header is the first one, vs the last. This may seem a bit random, but in a 600MB log with many requests, only requests with `Content-Length: 0` have that header come last. All others see the header come first.

I'm going to try to reproduce this with 1.52, as well as with `rclone restic serve`, see if I can spot anything weird.

**Comment by ncw** [comment]
> Thus, it doesn't look like that code is setting the wrong size -- otherwise we'd see it in the message. So either the wrong size gets set elsewhere in the server, or it's misinterpreted by restic.

Well spotted.



> One extra weird detail, that may or may not be relevant is the actual position of the header in the logs...

Hmm... I suspect that is relevant. Those http dumps you see aren't what is on the wire - http2 is a binary format, so they've been reconstructed. I suspect the 0 at the end might mean the length was missing.

It would be possible to catch the http2 traffic between restic and rclone by running a stdin/stdout duplicator.. I can't think of one off the shelf but it would be easy to write...

**Comment by ncw** [comment]
It would be a good step in debugging to set up rclone as a standalone restic rest server and try it like that to see if it has the same problem.

**Comment by ivoanjo** [comment]
I've been picking this up on and off, and it occurred to me to try something like this:

```diff
diff --git a/cmd/serve/httplib/serve/serve.go b/cmd/serve/httplib/serve/serve.go
index 89c4f696d..77483d673 100644
--- a/cmd/serve/httplib/serve/serve.go
+++ b/cmd/serve/httplib/serve/serve.go
@@ -4,6 +4,7 @@ package serve
 import (
        "fmt"
        "io"
+       "io/ioutil"
        "net/http"
        "path"
        "strconv"
@@ -83,6 +84,14 @@ func Object(w http.ResponseWriter, r *http.Request, o fs.Object) {
 
        w.WriteHeader(code)
 
+       _, err = ioutil.ReadFile("/tmp/simulate-failure")
+
+    if err == nil {
+        fs.Errorf(o, "Simulating failure!")
+        return
+    }
+
        n, err := io.Copy(w, in)
        if err != nil {
                fs.Errorf(o, "Didn't finish writing GET request (wrote %d/%d bytes): %v", n, size, err)
```

and it does trigger the issue!

This can be used to reproduce the issue on-demand: I can start a restic check, and then simulate the failure as needed at any point by creating or deleting the `/tmp/simulate-failure` file.

**Comment by ncw** [comment]
> This can be used to reproduce the issue on-demand: I can start a restic check, and then simulate the failure as needed at any point by creating or deleting the `/tmp/simulate-failure` file.

Do you mean it gives a `Content-Length: 0` when you look at the http debug from restic?

Could you show the restic log produced with that?

**Comment by ivoanjo** [comment]
> Do you mean it gives a Content-Length: 0 when you look at the http debug from restic?

It does!

Sharing the log:

* Restic stdout:

```
+ restic -o rclone.program=/home/knuckles/projects/rclone/rclone check --read-data --verbose --verbose --verbose --verbose --verbose --verbose
debug log file restic-debug-2020-06-08T20:50+01:00.log
debug enabled
using temporary cache in /tmp/restic-check-cache-170576950
repository 55558a0c opened successfully, password is correct
created new cache in /tmp/restic-check-cache-170576950
create exclusive lock for repository
load indexes
check all packs
check snapshots, trees and blobs
read all data
<-- during the data check, I create the simulate-error file
rclone: 2020/06/08 20:53:39 ERROR : data/d6/d63635ae60593073b1286715c1130e6e1e42a822117448a599f76b8db7449c99: Simulating failure!
Pack ID does not match, want d63635ae, got e3b0c442
rclone: 2020/06/08 20:53:39 ERROR : data/dc/dce568bab4c2df47849dd95a2d1bf721adf73cc3669f024d7ac3421f72870de1: Simulating failure!
Pack ID does not match, want dce568ba, got e3b0c442
rclone: 2020/06/08 20:53:40 ERROR : data/5c/5c039941304864142f74af969fbd0ee8d135cdecaa37c322e1b8302dfeaa78f4: Simulating failure!
Pack ID does not match, want 5c039941, got e3b0c442
rclone: 2020/06/08 20:53:40 ERROR : data/e7/e771ead093a0cc283f1e08944fa111eeaa7012a7d6991d95c6aaea5cb96329cc: Simulating failure!
Pack ID does not match, want e771ead0, got e3b0c442
rclone: 2020/06/08 20:53:40 ERROR : data/11/11470a25b256e0aedfc5b039094d4e62d33ca406f69d668f96dc0591e227a206: Simulating failure!
```

and in the debug log:

```
2020/06/08 20:53:42 debug/round_tripper_debug.go:77	debug.loggingRoundTripper.RoundTrip	245	------------  HTTP REQUEST -----------
GET /data/a7af93d1355e1e9bb6bfea0cbffe7f7325f07252c042afb9c1030a883d721315 HTTP/1.1
Host: localhost
User-Agent: Go-http-client/1.1
Accept: application/vnd.x.restic.rest.v2
Range: bytes=0-


2020/06/08 20:53:42 debug/round_tripper_debug.go:90	debug.loggingRoundTripper.RoundTrip	249	------------  HTTP RESPONSE ----------
HTTP/2.0 206 Partial Content
Accept-Ranges: bytes
Content-Range: bytes 0-7960901/7960902
Date: Mon, 08 Jun 2020 19:53:42 GMT
Server: rclone/v1.52.0-DEV-simulate-failure
Content-Length: 0


2020/06/08 20:53:42 checker/checker.go:712	checker.checkPack	249	hash for pack 9a918f2bb0eeb3e25039a9c53132f106d81dad6efedc10ffe2a110d3a6a1924f is e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
2020/06/08 20:53:42 checker/checker.go:715	checker.checkPack	249	Pack ID does not match, want 9a918f2bb0eeb3e25039a9c53132f106d81dad6efedc10ffe2a110d3a6a1924f, got e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
2020/06/08 20:53:42 checker/checker.go:699	checker.checkPack	249	checking pack 5fb9a3e6f53e3491bcb45c795ea08f15a69e9ec611834bc6f50d2773da80c636
2020/06/08 20:53:42 cache/backend.go:196	cache.(*Backend).Load	249	Load(<data/5fb9a3e6f5>, 0, 0): delegating to backend
2020/06/08 20:53:42 rest/rest.go:180	rest.(*Backend).openReader	249	Load <data/5fb9a3e6f5>, length 0, offset 0
2020/06/08 20:53:42 rest/rest.go:204	rest.(*Backend).openReader	249	Load(<data/5fb9a3e6f5>) send range bytes=0-
2020/06/08 20:53:42 debug/round_tripper_debug.go:77	debug.loggingRoundTripper.RoundTrip	249	------------  HTTP REQUEST -----------
GET /data/5fb9a3e6f53e3491bcb45c795ea08f15a69e9ec611834bc6f50d2773da80c636 HTTP/1.1
Host: localhost
User-Agent: Go-http-client/1.1
Accept: application/vnd.x.restic.rest.v2
Range: bytes=0-


2020/06/08 20:53:42 debug/round_tripper_debug.go:90	debug.loggingRoundTripper.RoundTrip	249	------------  HTTP RESPONSE ----------
HTTP/2.0 206 Partial Content
Accept-Ranges: bytes
Content-Range: bytes 0-4437486/4437487
Date: Mon, 08 Jun 2020 19:53:42 GMT
Server: rclone/v1.52.0-DEV-simulate-failure
Content-Length: 0


2020/06/08 20:53:42 checker/checker.go:712	checker.checkPack	249	hash for pack 5fb9a3e6f53e3491bcb45c795ea08f15a69e9ec611834bc6f50d2773da80c636 is e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
2020/06/08 20:53:42 checker/checker.go:715	checker.checkPack	249	Pack ID does not match, want 5fb9a3e6f53e3491bcb45c795ea08f15a69e9ec611834bc6f50d2773da80c636, got e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
2020/06/08 20:53:42 checker/checker.go:699	checker.checkPack	249	checking pack f304e611855323d5e90ff339d5f920b675557920b9f8e3c0903f943fe36d9e2a
2020/06/08 20:53:42 cache/backend.go:196	cache.(*Backend).Load	249	Load(<data/f304e61185>, 0, 0): delegating to backend
2020/06/08 20:53:42 rest/rest.go:180	rest.(*Backend).openReader	249	Load <data/f304e61185>, length 0, offset 0
2020/06/08 20:53:42 rest/rest.go:204	rest.(*Backend).openReader	249	Load(<data/f304e61185>) send range bytes=0-
2020/06/08 20:53:42 debug/round_tripper_debug.go:77	debug.loggingRoundTripper.RoundTrip	249	------------  HTTP REQUEST -----------
GET /data/f304e611855323d5e90ff339d5f920b675557920b9f8e3c0903f943fe36d9e2a HTTP/1.1
Host: localhost
User-Agent: Go-http-client/1.1
Accept: application/vnd.x.restic.rest.v2
Range: bytes=0-


2020/06/08 20:53:42 debug/round_tripper_debug.go:90	debug.loggingRoundTripper.RoundTrip	246	------------  HTTP RESPONSE ----------
HTTP/2.0 206 Partial Content
Accept-Ranges: bytes
Content-Range: bytes 0-6375177/6375178
Date: Mon, 08 Jun 2020 19:53:42 GMT
Server: rclone/v1.52.0-DEV-simulate-failure
Content-Length: 0


2020/06/08 20:53:42 checker/checker.go:712	checker.checkPack	246	hash for pack ab79255e74a6a82599da0420ba3e81d608a713b48055b2a97ad204e1cb25dc72 is e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
2020/06/08 20:53:42 checker/checker.go:715	checker.checkPack	246	Pack ID does not match, want ab79255e74a6a82599da0420ba3e81d608a713b48055b2a97ad204e1cb25dc72, got e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
2020/06/08 20:53:42 checker/checker.go:699	checker.checkPack	246	checking pack f5101d9d80deb97e21e02143ff8732f07e078921d881d0dae1952cdc184d2a04
2020/06/08 20:53:42 cache/backend.go:196	cache.(*Backend).Load	246	Load(<data/f5101d9d80>, 0, 0): delegating to backend
2020/06/08 20:53:42 rest/rest.go:180	rest.(*Backend).openReader	246	Load <data/f5101d9d80>, length 0, offset 0
2020/06/08 20:53:42 rest/rest.go:204	rest.(*Backend).openReader	246	Load(<data/f5101d9d80>) send range bytes=0-
2020/06/08 20:53:42 debug/round_tripper_debug.go:77	debug.loggingRoundTripper.RoundTrip	246	------------  HTTP REQUEST -----------
GET /data/f5101d9d80deb97e21e02143ff8732f07e078921d881d0dae1952cdc184d2a04 HTTP/1.1
Host: localhost
User-Agent: Go-http-client/1.1
Accept: application/vnd.x.restic.rest.v2
Range: bytes=0-


2020/06/08 20:53:42 debug/round_tripper_debug.go:90	debug.loggingRoundTripper.RoundTrip	247	------------  HTTP RESPONSE ----------
HTTP/2.0 206 Partial Content
Accept-Ranges: bytes
Content-Range: bytes 0-5685271/5685272
Date: Mon, 08 Jun 2020 19:53:42 GMT
Server: rclone/v1.52.0-DEV-simulate-failure
Content-Length: 0


2020/06/08 20:53:42 checker/checker.go:712	checker.checkPack	247	hash for pack bd09b91842da34118612a071efb5eebcf1d443be326563e38214f08ef062b6fb is e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
2020/06/08 20:53:42 checker/checker.go:715	checker.checkPack	247	Pack ID does not match, want bd09b91842da34118612a071efb5eebcf1d443be326563e38214f08ef062b6fb, got e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
```

Let me know if I can provide any more info!

**Comment by ncw** [comment]
So I think that means rclone isn't sending the data. The original error was " http2: stream closed" so I think your test is simulating that quite well, but I don't think we know why we get that error.

I suspect it is something to do with the whole http2 over stdin thing.

I think there are some http2 debugging flags, maybe you could see what you can get out of those?

https://stackoverflow.com/questions/36513391/is-there-any-way-to-enable-detailed-tracing-of-http2-requests-and-responses-in-g

**Comment by ivoanjo** [comment]
Log from running with debug flags with `GODEBUG=http2debug=1`:

```
rclone: 2020/06/09 08:20:47 http2: server read frame HEADERS flags=END_STREAM|END_HEADERS stream=557 len=58
rclone: 2020/06/09 08:20:47 http2: Transport failed to get client conn for www.jottacloud.com:443: http2: no cached connection was available
rclone: 2020/06/09 08:20:47 ERROR : data/c6/c61c577d4ce1e41da9cf286051cda6ea1339ae510400280b242a6ec3ad1a1814: Simulating failure!
rclone: 2020/06/09 08:20:47 http2: server encoding header ":status" = "206"
rclone: 2020/06/09 08:20:47 http2: server encoding header "accept-ranges" = "bytes"
rclone: 2020/06/09 08:20:47 http2: server encoding header "content-range" = "bytes 0-5062415/5062416"
rclone: 2020/06/09 08:20:47 http2: server encoding header "server" = "rclone/v1.52.0-DEV-simulate-failure"
rclone: 2020/06/09 08:20:47 http2: server encoding header "content-length" = "5062416"
rclone: 2020/06/09 08:20:47 http2: server encoding header "date" = "Tue, 09 Jun 2020 07:20:47 GMT"
Pack ID does not match, want c61c577d, got e3b0c442
rclone: 2020/06/09 08:20:47 http2: server read frame HEADERS flags=END_STREAM|END_HEADERS stream=559 len=58
rclone: 2020/06/09 08:20:47 http2: Transport failed to get client conn for www.jottacloud.com:443: http2: no cached connection was available
rclone: 2020/06/09 08:20:47 http2: Transport failed to get client conn for www.jottacloud.com:443: http2: no cached connection was available
rclone: 2020/06/09 08:20:47 ERROR : data/41/41d798cd47eaa64406fa8f446ad431d3c02ea8dede6bc7ebb2d4e8685879e059: Simulating failure!
rclone: 2020/06/09 08:20:47 http2: server encoding header ":status" = "206"
rclone: 2020/06/09 08:20:47 http2: server encoding header "accept-ranges" = "bytes"
rclone: 2020/06/09 08:20:47 http2: server encoding header "content-range" = "bytes 0-5798732/5798733"
rclone: 2020/06/09 08:20:47 http2: server encoding header "server" = "rclone/v1.52.0-DEV-simulate-failure"
rclone: 2020/06/09 08:20:47 http2: server encoding header "content-length" = "5798733"
rclone: 2020/06/09 08:20:47 http2: server encoding header "date" = "Tue, 09 Jun 2020 07:20:47 GMT"
Pack ID does not match, want 41d798cd, got e3b0c442
```
(the restic debug log still says it's getting `Content-Length: 0`)

**Comment by ncw** [comment]
> Log from running with debug flags with `GODEBUG=http2debug=1`:

That is a good start, but nothing I can see that is out of the ordinary.

Any chance you could capture a failure without using your failure injector?

**Comment by buengese** [comment]
Didn't notice before that there was an open issue for this. I had been looking into this a few months ago and started looking back into this today so I may be able to add some information.

- This really only happens if you use the http2 over stdio connection. So running on a normal tcp socket does serve as workaround
- This does not appear to happen if you serve a local remote which makes me believe the problem isn't actually the http2 over stdio by itself but probably caused by the upstream http connection to the remote.

**Comment by ncw** [comment]
> * This really only happens if you use the http2 over stdio connection. So running `serve restic` on a normal tcp socket does work as workaround

That is good to know. The `http2 over stdio` is a bit magic but it is very simple so I can't see what can be wrong with rclone's code.

Did you try http2 over TCP? That would probably require a bit of messing about with TLS certificates or some code hacking...



> * This does not appear to happen if you serve a local remote which makes me believe the problem isn't actually the http2 over stdio by itself but probably caused by some weird interaction with the upstream http connection to the remote.

I think what triggers it is some kind of error that is only detected after we've sent the HTTP headers - something that is returned by the `io.Copy`.

**Comment by ivandeex** [comment]
This happens because restic's `Load()` does not wait for EOF on the body stream from rclone after
reading full body of a partial object. Since restic and rclone internally talk HTTP/2 on their stdio pipe, this sometimes makes Golang's HTTP/2 client send back a `stream reset` frame to rclone.
Please note that it happens *after* rclone has completed sending partial object and does not cause data corruption.

The issue was investigated here: https://github.com/ivandeex/rclone/tree/stream-reset/stream-reset#premature-stream-reset-between-restic-and-rclone
and can be solved by  making restic drain rclone response body till EOF before calling `Close` as proposed in https://github.com/restic/restic/pull/3014

The drain approach was once discussed in https://github.com/restic/restic/issues/1561#issuecomment-373118890 but was not implemented.

A similar issue (*RST_STREAM + CANCEL sent by HTTP2 client to server after reading the DATA frame of a response*) was reported against Golang HTTP/2 library in https://github.com/golang/go/issues/41570#issuecomment-706765842
Even if Golang authors confirm it as a bug, it will not be fixed before the go v1.17
It can be reproduced by running a simple test https://github.com/ivandeex/rclone/blob/stream-reset/stream-reset/http2test.go which does not require restic or rclone:
```
go run http2test.go http1 nodrain 10
```
runs successfully
```
go run http2test.go http2 nodrain 10
```
prints error messages
```
go run http2test.go http2 drain 10
```
is successful again.

**Comment by ncw** [comment]
Top notch debugging @ivandeex :-)

I commented on the restic PR.

I bet rclone makes this mistake in various places too...

**Comment by fd0** [comment]
PR https://github.com/restic/restic/pull/3014 in restic is merged, thanks a lot @ivandeex!

**Comment by ivandeex** [comment]
@bkmeneguello @ivoanjo @azurefreecovid @tipruzs 

The fix will be released with restic 0.12
The beta based on 0.10 with fix applied is here https://github.com/ivandeex/restic/releases
You are welcome to give it a try

**Comment by ivandeex** [comment]
Closing per above.
